src/t_stream.c

```diff
@@ -5120,3 +5383,392 @@ void handleClaimableStreamEntries(void) {
+                while (producer->idmp_head != NULL) {
+                    idmpEntry *entry = producer->idmp_head;
+                    if (entry->id.ms <= expire_time) {
+                        /* Remove from dict */
+                        dictDelete(producer->idmp_dict, entry);
+                        /* Remove from linked list head */
+                        producer->idmp_head = entry->next;
+                        if (producer->idmp_head == NULL) {
+                            producer->idmp_tail = NULL;
+                        }
+                        /* Free the entry */
+                        idmpEntryFree(entry, &s->alloc_size);
+                    } else {
+                        break;
+                    }
+                }
+
+                if (producer->idmp_head != NULL) {
+                    has_entries = 1;
+                }
+            }
+            raxStop(&ri);
+
+            if (!has_entries) {
+                dictDelete(db->stream_idmp_keys, key);
+                continue;
+            }
+        }
+        dictResetIterator(&di);
+    }
+}
+
+/* Hash field-value pairs using XXH3_128bits for AUTOIDMP. The function
+ * takes an array of robj pointers in 'argv' representing field-value pairs
+ * (field1, value1, field2, value2, ...) and 'numfields' indicating the number
+ * of pairs (not the array length). Each field-value pair is hashed using
+ * streaming XXH3_128bits, and the resulting pair hashes are XORed together
+ * to produce a final 128-bit hash. Returns a zero hash {0, 0} on error. */
+static XXH128_hash_t createIdempotencyHash(robj **argv, int64_t numfields) {
+    XXH128_hash_t hash_result = {0, 0};
+    XXH3_state_t* state = XXH3_createState();
+    if (state == NULL) return hash_result;
+    
+    char llbuf[LONG_STR_SIZE];
+    XXH_errorcode err;
+    
+    /* Process each field-value pair */
+    for (int64_t i = 0; i < numfields; i++) {
+        robj *field = argv[i * 2];
+        robj *value = argv[i * 2 + 1];
+        
+        /* Initialize hash state for this pair */
+        err = XXH3_128bits_reset(state);
+        if (err != XXH_OK) goto cleanup;
+        
+        /* Hash the field */
+        long field_len;
+        unsigned char *field_data = getObjectReadOnlyString(field, &field_len, llbuf);
+        err = XXH3_128bits_update(state, field_data, field_len);
+        if (err != XXH_OK) goto cleanup;
+        
+        /* Hash the value */
+        long value_len;
+        unsigned char *value_data = getObjectReadOnlyString(value, &value_len, llbuf);
+        err = XXH3_128bits_update(state, value_data, value_len);
+        if (err != XXH_OK) goto cleanup;
+        
+        /* Get the hash for this pair */
+        XXH128_hash_t pair_hash = XXH3_128bits_digest(state);
+        
+        /* XOR with accumulated result */
+        hash_result.low64 ^= pair_hash.low64;
+        hash_result.high64 ^= pair_hash.high64;
```

Comment:
is the hash of `a 1 b 2` the same as `a 2 b 1`?
add a hash multiplier for it?

```suggestion
        hash_result.low64 = (hash_result.low64 * 31) ^ pair_hash.low64;
        hash_result.high64 = (hash_result.high64 * 31) ^ pair_hash.high64;
```

Reply:
The function must generate the same hash regardless of the order of the field-value pairs.

Reply:
I think what you said is that `a 1 b 2` and `b 2 a 1` should generate the same hash, but should `a 1 b 2` and `a 2 b 1` also be the same?

Reply:
`a 1 b 2` and `a 2 b 1` produce different hash.
`a 1 b 2` and `b 2 a 1` produce the same hash.

Reply:
i misunderstood that the faield and value would be a fixed hash.